File: fgraph-gui/layout.py
"""3D radial tree layout: same scheme as the terminal version's layout.rs.

Each node gets an angular slice proportional to its subtree size, radius
grows with depth, and elevation is nudged per-node (stable hash of the path)
so the tree blooms into a dome instead of sitting flat on one plane.
"""
from __future__ import annotations

import hashlib
import math

from scan import Graph

RADIUS_PER_DEPTH = 3.0


def _stable_unit(path) -> float:
    digest = hashlib.md5(str(path).encode("utf-8", "ignore")).digest()
    return int.from_bytes(digest[:4], "big") / 0xFFFFFFFF


def layout(graph: Graph) -> None:
    _assign(graph, graph.root, 0.0, math.tau, 0.0)


def _assign(graph: Graph, idx: int, angle_start: float, angle_end: float, elevation_bias: float):
    node = graph.nodes[idx]
    angle_mid = (angle_start + angle_end) * 0.5
    radius = node.depth * RADIUS_PER_DEPTH

    jitter = (_stable_unit(node.path) - 0.5) * 0.6
    elevation = max(-1.4, min(1.4, elevation_bias + jitter))
    polar = math.pi / 2 - elevation

    x = radius * math.sin(polar) * math.cos(angle_mid)
    y = radius * math.cos(polar)
    z = radius * math.sin(polar) * math.sin(angle_mid)
    node.pos = (x, y, z)

    children = node.children
    if not children:
        return

    total_leaves = sum(graph.nodes[c].leaf_count for c in children) or 1
    cursor = angle_start
    span = angle_end - angle_start
    for c in children:
        share = graph.nodes[c].leaf_count / total_leaves
        child_start = cursor
        child_end = cursor + span * share
        child_bias = elevation_bias + (_stable_unit(graph.nodes[c].path) - 0.5) * 0.5
        _assign(graph, c, child_start, child_end, child_bias)
        cursor = child_end

File: fgraph/src/layout.rs
use crate::scan::Graph;
use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};

const RADIUS_PER_DEPTH: f32 = 6.0;

/// Radial tree layout, projected onto a sphere-ish 3D shape:
/// each node gets an angular slice proportional to its subtree size (leaf_count),
/// radius grows with depth, and elevation is nudged per-node (stable hash of the
/// path) so the tree reads as a 3D dome/bloom instead of a flat disk when rotated.
pub fn layout(graph: &mut Graph) {
    assign(graph, graph.root, 0.0, std::f32::consts::TAU, 0.0);
}

fn stable_unit(path: &std::path::Path) -> f32 {
    let mut hasher = DefaultHasher::new();
    path.hash(&mut hasher);
    (hasher.finish() % 10_000) as f32 / 10_000.0
}

fn assign(graph: &mut Graph, idx: usize, angle_start: f32, angle_end: f32, elevation_bias: f32) {
    let depth = graph.nodes[idx].depth;
    let angle_mid = (angle_start + angle_end) * 0.5;
    let radius = depth as f32 * RADIUS_PER_DEPTH;

    let jitter = (stable_unit(&graph.nodes[idx].path) - 0.5) * 0.6;
    let elevation = elevation_bias + jitter;
    let polar = std::f32::consts::FRAC_PI_2 - elevation.clamp(-1.4, 1.4);

    let x = radius * polar.sin() * angle_mid.cos();
    let y = radius * polar.cos();
    let z = radius * polar.sin() * angle_mid.sin();
    graph.nodes[idx].pos = [x, y, z];

    let children = graph.nodes[idx].children.clone();
    if children.is_empty() {
        return;
    }
    let total_leaves: f32 = children
        .iter()
        .map(|c| graph.nodes[*c].leaf_count as f32)
        .sum();
    let mut cursor = angle_start;
    let span = angle_end - angle_start;
    for c in children {
        let share = graph.nodes[c].leaf_count as f32 / total_leaves.max(1.0);
        let child_start = cursor;
        let child_end = cursor + span * share;
        let child_bias = elevation_bias + (stable_unit(&graph.nodes[c].path) - 0.5) * 0.5;
        assign(graph, c, child_start, child_end, child_bias);
        cursor = child_end;
    }
}
